src/ddmd/cond.d

```diff
@@ -64,6 +67,227 @@ extern (C++) abstract class Condition : RootObject
 }
 
 /***********************************************************
+ * This class implements common functionality for StaticForeachDeclaration and StaticForeachStatement
+ */
+
+immutable StaticForeach_tupleFieldName = "tuple"; // used in lowering
+extern (C++) final class StaticForeach : RootObject
+{
+    Loc loc;
+
+    ForeachStatement aggrfe;
+    ForeachRangeStatement rangefe;
+
+    bool needExpansion = false; // need to expand a tuple into multiple variables
+
+    final extern (D) this(Loc loc,ForeachStatement aggrfe,ForeachRangeStatement rangefe)
+    in
+    {
+        assert(!!aggrfe^!!rangefe);
+    }
+    body
+    {
+        this.loc=loc;
+        this.aggrfe=aggrfe;
+        this.rangefe=rangefe;
+    }
+
+    StaticForeach syntaxCopy(){
+        return new StaticForeach(
+            loc,
+            aggrfe?cast(ForeachStatement)aggrfe.syntaxCopy():null,
+            rangefe?cast(ForeachRangeStatement)rangefe.syntaxCopy():null
+        );
+    }
+
+    private extern(D) void lowerArrayAggregate(Scope* sc)
+    {
+        auto aggr = aggrfe.aggr;
+        Expression el = new ArrayLengthExp(aggr.loc, aggr);
+        sc = sc.startCTFE();
+        el = el.semantic(sc);
+        sc = sc.endCTFE();
+        el = el.optimize(WANTvalue);
+        el = el.ctfeInterpret();
+        if (el.op == TOKint64){
+            dinteger_t length = el.toInteger();
+            auto es = new Expressions();
+            foreach(i;0..length)
+            {
+                auto index = new IntegerExp(loc, i, Type.tsize_t);
+                auto value = new IndexExp(aggr.loc, aggr, index);
+                es.push(value);
+            }
+            aggrfe.aggr = new TupleExp(aggr.loc, es);
+            aggrfe.aggr = aggrfe.aggr.semantic(sc);
+            aggrfe.aggr = aggrfe.aggr.optimize(WANTvalue);
+        }
+        else
+        {
+            aggrfe.aggr = new ErrorExp();
+            return;
+        }
+    }
+
+    private extern(D) Expression evaluate(Loc loc,Statement s)
```

Comment:
Evaluate? The function body does not evaluate anything - it appears to do a rewrite. Please document.